### PR TITLE
Add Concurrency Kit checks and define setting to configure.ac

### LINKED ISSUE #221: concurrencykit is not checked for during configure
src/rtapi/rtapi_atomics.h and some other places have support for concurrencykit.org:

```
 26 #include "config.h" // HAVE_CK
 31 #ifdef HAVE_CK
```

However there is no configure rule for HAVE_CK.

```
$ git grep HAVE_CK
src/hal/icomp-example/pin.hh:#ifdef HAVE_CK
src/hal/lib/hal_accessor.h:#include "config.h"  // HAVE_CK
src/hal/lib/hal_accessor.h:#ifdef HAVE_CK
src/regressions/check-tests/check_tests.c:#ifdef HAVE_CK
src/regressions/check-tests/check_tests.c:#ifdef HAVE_CK
src/rtapi/rtapi.h:#ifdef HAVE_CK
src/rtapi/rtapi_atomics.h:#include "config.h" // HAVE_CK
src/rtapi/rtapi_atomics.h:#ifdef HAVE_CK
src/rtapi/rtapi_atomics.h:#if !defined(HAVE_CK) || !defined(CK_F_PR_LOAD_64)
src/rtapi/rtapi_atomics.h:#if !defined(HAVE_CK) || !defined(CK_F_PR_STORE_64)
src/rtapi/rtapi_atomics.h:#if !defined(HAVE_CK) || !defined(CK_F_PR_INC_64)
src/rtapi/rtapi_atomics.h:#if !defined(HAVE_CK) || !defined(CK_F_PR_CAS_64)
src/rtapi/rtapi_atomics.h:#if !defined(HAVE_CK)
src/rtapi/rtapi_atomics.h:#endif  // !defined(HAVE_CK)
src/rtapi/rtapi_heap_private.h:#ifdef HAVE_CK
src/rtapi/rtapi_int.h:#ifdef HAVE_CK
src/rtapi/rtapi_msgd.cc:#ifdef HAVE_CK
```

The default without ck will be expensive full memory barriers for all of rmb/wmb.

machinekit-cnc seems to have a similar situation.

### PR DESCRIPTION
**Pull request by ArcEye**
Part of the multicore work which is missing for some reason.

Fixes #221 

Signed-off-by: Mick <arceye@mgware.co.uk>

### CODE REVIEW

**Comment by ArcEye** [comment]
The checks are made, but because the dockers do not contain `libck` the build is not changed at present
```
checking for CK... no
not using concurrencykit
```
Anyone building a RIP with libck installed will find it working however.

Next stage can be done once dockers updated

**Comment by ArcEye** [comment]
Hold this one for now.

Thinking about packages, the existence of libck is going to have to be a compulsory dependency, if the packages are being built linked against it.

In RIP builds, which is all I use, it does not matter, it will either be used or not depending upon whether installed on the system.

I will add a failure path to the configure.ac if it is not installed and add the `debian/control*` lines to enforce usage for builds and installs and it can be committed as one lump.

It will however then fail until the dockers are updated, which is what I was looking to avoid with this half way house PR

**Comment by ArcEye** [comment]
This is working as it should, configure fails because concurrencykit is not in the dockers
```
checking for CK... no
configure: error:  concurrencykit not found - please install package libck0
```

Once @zultron has regenerated the dockers, this can be re-tested and merged.

File: debian/control.d/control.buster.in
Comment (zultron) [inline debian/control.d/control.buster.in]: There's no `libck0` package in Buster.

File: debian/control.d/control.in
Comment (zultron) [inline debian/control.d/control.in]: There's no `libck-dev` package in Buster, so this needs to be moved into `debian/configure` and added through the `BUILD_DEPS` variable.

File: debian/control.d/control.jessie.in
Comment (zultron) [inline debian/control.d/control.jessie.in]: `dh_shlibdeps` should be able to pick this up automatically, and this can be dropped.  Same for the hunk in `control.stretch.in`.

**Review by zultron — CHANGES_REQUESTED** [review]
Thanks for the PR.  Buster doesn't have a `libck-dev` package at all, AFAICT (but Sid does?!?), so that needs fixing.  The other comment about `libck0` being picked up by `dh_shlibdeps` is just a nit.

File: debian/control.d/control.buster.in
Comment (ArcEye) [reply debian/control.d/control.buster.in]: Don't know what is going on with Debian at present.

This situation is the same for quite a few packages at present, if it is in sid, why don't they just backport it, instead of leaving Buster with no package?

I am actually using sid presently so did not even think to check if Buster was covered.

Does your `/etc/apt/sources.list.d` in the dockers have a `machinekit.list` file?
I am very tempted to just build a `libck0 / libck-dev` packages for Buster and host it in the repo, rather than have the stupid situation where people running the most up to date distro, cannot use the ck features.

Comment (zultron) [reply debian/control.d/control.buster.in]: Yes, this is dumb.  I'll ask around the Debian community at some point.

The mk-cross-builder images do configure the MK apt repository.  If you add that package, let me know and I'll update the images.

Did you see my comment in #221, `libck` is also not built for Jessie i386?

File: src/configure.ac
Comment (zultron) [inline src/configure.ac]: Nit:  This line is unnecessary:
```
checking for CK... yes
concurrencykit found - configuring to use ck
```

File: src/configure.ac
Comment (zultron) [inline src/configure.ac]: This will cause the build to exit if `libck` isn't available.  Since `libck` isn't shipped for Jessie i386, can this line just be dropped, and `libck` be optional?

**Review by zultron — CHANGES_REQUESTED** [review]
Putting `libck` packages in the MK repo for Buster is worth the effort, but Jessie i386 doesn't seem like it.  I think `libck` should be made optional by removing the relevant line from this PR to work around this.

File: src/configure.ac
Comment (zultron) [inline src/configure.ac]: Builds fail on Jessie, where `libck` is version 0.3.5.

File: src/configure.ac
Comment (zultron) [reply src/configure.ac]: Ah ha, turns out the version of `libck` in Jessie is 0.3.5, whereas this patch requires `PKG_CHECK_MODULES([CK], [ck >= 0.4.6], [...])`.  Backing this down to 0.3.5 and building in Jessie prints lots of errors like this:

```
hal/lib/hal_accessor.h:258:230: error: ‘ck_pr_load_double’ was not declared in this scope
```

So I think `libck` must be optional at least as long as we support Jessie.

**Comment by zultron** [comment]
The Jenkins tests pass over in #224 with the new images and a commit I added to make `libck` optional, since Jessie ships v. 0.3.5, incompatible with the API used in MK.

However, Jenkins seems to test only `amd64_9` and `armhf_8`.  Because of the complete absence of `libck` for Jessie i386 and Buster (all arches), package builds will fail.  I added a second, untested commit to fix up dependencies, following the comments I put in the review here.

**Comment by ArcEye** [comment]
I won't have any time for about a week to finish this.

The reason I changed it to compulsory, was that if the packages were built, linked against `libck.so`, it could not be optional as to whether that lib was present on the host machine.

If Jessie does not have the correct ck version or an i386 libck0, then I suggest we just scrap ck for Jessie (which will be EOL next year anyway).

I have back ported the sid libck0 and libck-dev packages into the machinekit Buster
eg. http://deb.machinekit.io/debian/dists/buster/main/binary-amd64/Packages
Not had a chance to test if they run OK on Buster, but think that given the other lib versions and compiler versions they probably would.

**Comment by ArcEye** [comment]
I have just pulled the packages from machinekit repo to a Buster partition, then built and run https://github.com/arceye/machinekit-hal/concurrency-kit-checks branch

So the packages in sid work fine with Buster, as I thought they would, which makes the omission from the Debian repos even stranger.

**Comment by zultron** [comment]
@ArcEye Looking good, thanks for the Buster packages.  The builds are [passing][1] on the latest images that include `libck` in Stretch (from upstream) and Buster (from your rebuild), and don't include `libck` in Jessie.

In order to make that all work on the MK side, I needed a [couple of extra changes][2] on top of yours.

The top commit in that branch is just to point the build at the [new test images][3] temporarily.  Once we decide those images are OK I can promote them to the (de-facto) ["official" image repo][4].  Then you can pull whatever changes from #224, confirm that the build passes one last time, and review & merge.

[1]: https://travis-ci.org/zultron/machinekit-hal/builds/549821422
[2]: https://github.com/zultron/machinekit-hal/commits/concurrency-kit-checks
[3]: https://hub.docker.com/r/zultron/mk-cross-builder/tags/
[4]: https://hub.docker.com/r/dovetailautomata/mk-cross-builder/tags

**Comment by ArcEye** [comment]
@zultron I have pulled your last 3 commits into my branch and the result is testing now.

If this passes, I will revert the docker address and wait to hear from you that the dockers are updated and this can be merged.

**Comment by ArcEye** [comment]
That worked
```
+ scripts/build_docker -t amd64_9 -c test
amd64_9: Pulling from zultron/mk-cross-builder
```
```
checking for CK... yes
```

@zultron  - let me know when main dockers are updated and when I next get a window, will change this PR back to the old docker specs and get it merged.

**Comment by zultron** [comment]
Yep, my Travis CI config builds all distro+arch combos, and I verified `checking for CK... (yes|no)` as appropriate for each in the [recent build][1].

I submitted and merged :P the image updates in Dovetail-Automata/mk-cross-builder#6.  If I didn't do anything dumb, we should see newly-built images show up [here][2] in the next few hours.

[1]: https://travis-ci.org/zultron/machinekit-hal/builds/549821422
[2]:  https://hub.docker.com/r/dovetailautomata/mk-cross-builder/tags

**Comment by zultron** [comment]
Before merging:

- Don't forget to pop off the top commit, 6be3dccd "FIXME CI: Build with test Docker images"
- I recommend kicking off another Jenkins build just to test the new Docker images one last time

Also, last time, stale images on Jenkins caused problems with free disk space, if that's still a concern.

**Comment by ArcEye** [comment]
Thanks @zultron 
I will test locally when I get a mo, then delete all the dockers from the server, so that it pulls the new ones before each build.

**Comment by ArcEye** [comment]
I have deleted all the old dockers, plus @zultron's testing ones for #224  
In the process of pulling the new dockers, then I will do a forced update with current HEAD removed to test the build.

**Comment by ArcEye** [comment]
OK this now builds as it should - ready to merge.

**Comment by ArcEye** [comment]
@zultron and I have tested this to death, looks like everyone else is on holiday, merging.

**Pull request merged**
